### PR TITLE
Extract WorkspaceSettingsModalStackNavigator

### PR DESCRIPTION
**Pull request by hayata-suenaga**
### Details
There's a new requirement to implement protected routes. These are routes or screens that only a specific group of users should access. To do this, we need to conditionally add or remove screens from a navigator.

In our existing code, all stack navigators are created using the `createModalStackNavigator` function. But this function doesn't support conditional screens.

This proposed change redefines the `WorkspaceSettingsModalStackNavigator` without using the `createModalStackNavigator` function. This allows certain workspace settings-related screens to be made conditionally available.

Please note that this proposal doesn't implement the logic for conditionally registering certain screens. But it makes it easy to implement this kind of logic.

You can find further information on this change in [this Slack thread](https://expensify.slack.com/archives/C01GTK53T8Q/p1710982092242099).

### Fixed Issues
$ https://github.com/Expensify/App/issues/39087
PROPOSAL: N/A

### Tests / QA Test

1. Sign in to staging.expenisfy.com. Use a non-public domain if possible. If you need to use a public domain (e.g., google.com], append ?exitTo=inbox to the URL.
2. Navigate to Settings -> Workspaces -> Group -> New Workspace -> Collect. This creates a new Collect workspace for you. You do NOT need to enter any payment information.
3. On the screen for your newly created workspace, open the developer console. Paste and execute the following code snippet:
```
p = Policy.getCurrent();
p.policy.isPolicyExpenseChatEnabled = true;
p.save();
```
4. Sign in to New Expensify using the same account you used in step https://github.com/Expensify/App/pull/1.
5. Click on your avatar on the far right of the bottom tab -> Workspaces. The workspace you created should be listed. Click on it and then navigate to More Features.
6. Confirm that the screen you see matches the one shown below.
Screenshot 2024-03-27 at 11 43 19 AM
7. We want to check that all screens are accessible. Toggle all options listed on this page. Each time you toggle on an option, you should be navigated to the page for that option. After toggling on all options, the left-hand size of the screen should look something like this.
Screenshot 2024-03-27 at 11 45 50 AM

### Offline tests
N/A

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)

_I skipped the chrome on Android phone_

- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/Expensify/App/assets/98560306/54500bac-aadd-4a72-8a2b-fffab924ec67

_This is from Ad Hoc build_

</details>

<details>
<summary>Android: mWeb Chrome</summary>

N/A

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/Expensify/App/assets/98560306/258c0c63-8291-4325-9f0e-af9588f3d08d

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/Expensify/App/assets/98560306/f68741ad-efaa-4218-aa0a-6054c08e4426

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/98560306/31283834-5837-4555-9e34-3499d37e6aaa

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/Expensify/App/assets/98560306/7be76ba7-ea68-4407-92a4-47286ef6c377

</details>

### CODE REVIEW

**Comment by hayata-suenaga** [comment]
Regression #1

1. Open the ad hoc build on a web browser
2. Confirm that the `Hmm... it's not here` page is presented
3. Click the `Go back to home page` link
4. Confirm that the home page is presented
5. Reload the site
6. Confirm that the `Hmm... it's not here` page is presented again

**Review by situchan — APPROVED** [review]
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>


https://github.com/Expensify/App/assets/108292595/3675db0b-522d-4f88-8fbf-1a83249e69b3



</details>

<details>
<summary>Android: mWeb Chrome</summary>


https://github.com/Expensify/App/assets/108292595/c77e8dcf-b333-4e16-bdfe-a5d5b2f00612



</details>

<details>
<summary>iOS: Native</summary>


https://github.com/Expensify/App/assets/108292595/e0c3d063-b95e-49a3-ac3e-0f449732ed74



</details>

<details>
<summary>iOS: mWeb Safari</summary>


https://github.com/Expensify/App/assets/108292595/c085297a-0c07-404b-b804-2d8db79e036a



</details>

<details>
<summary>MacOS: Chrome / Safari</summary>


https://github.com/Expensify/App/assets/108292595/29927e3a-ca12-499d-9962-de37fe47afe6



</details>

<details>
<summary>MacOS: Desktop</summary>



https://github.com/Expensify/App/assets/108292595/82807bec-59d7-4e71-9abe-0de16745aacd


</details>

**Comment by situchan** [comment]
Please complete author checklist

**Comment by hayata-suenaga** [comment]
@roryabraham I'll work on the screen recordings after lunch. Meantime, could you review the code please? 🙇

**Comment by hayata-suenaga** [comment]
There remains an [[issue](https://github.com/Expensify/App/pull/39084#issuecomment-2023489833)](https://github.com/Expensify/App/pull/39084#issuecomment-2023489833) with the ad hoc build on the web, but the local build and builds on other platforms are not affected. Apart from this, the PR is ready for internal engineer review.

**Comment by situchan** [comment]
> There remains an [issue](https://github.com/Expensify/App/pull/39084#issuecomment-2023489833) with the ad hoc build on the web, but the local build and builds on other platforms are not affected. Apart from this, the PR is ready for internal engineer review.

yes, it's known issue happening on all adhoc builds

File: src/libs/Navigation/AppNavigator/WorkspaceSettingsModalStackNavigator.tsx [hunk 1]
```diff
@@ -0,0 +1,83 @@
+import {createStackNavigator} from '@react-navigation/stack';
+import React, {useMemo} from 'react';
+import useThemeStyles from '@hooks/useThemeStyles';
+import SCREENS from '@src/SCREENS';
+
+const StackNavigator = createStackNavigator();
+
+function WorkspaceSettingsModalStackNavigator() {
+    const styles = useThemeStyles();
+    const screenOptions = useMemo(() => ({cardStyle: styles.navigationScreenCardStyle, headerShown: false}), [styles]);
```
Comment (roryabraham) [inline src/libs/Navigation/AppNavigator/WorkspaceSettingsModalStackNavigator.tsx:10]: You're missing the `cardStyleInterpolator: CardStyleInterpolators.forHorizontalIOS` that's present in the `defaultSubRouteOptions` in `createModalStackNavigator`

**Review by roryabraham — CHANGES_REQUESTED** [review]
Also NAB but I suggest creating a hook to extract the common logic between `WorkspaceSettingsModalStackNavigator` and `createModalStackNavigator`

**Comment by hayata-suenaga** [comment]
Addressed Rory's PR reviews and retested with web

https://github.com/Expensify/App/assets/98560306/80a483a6-eafa-4d26-9b95-23541ed5c2d9

File: src/libs/Navigation/AppNavigator/WorkspaceSettingsModalStackNavigator.tsx (hunk 1 above)
Comment (hayata-suenaga) [reply src/libs/Navigation/AppNavigator/WorkspaceSettingsModalStackNavigator.tsx]: fixed 🙇

File: src/libs/Navigation/AppNavigator/ModalStackNavigators/useModalScreenOptions.ts [hunk 2]
```diff
@@ -0,0 +1,22 @@
+import type {StackNavigationOptions} from '@react-navigation/stack';
+import {CardStyleInterpolators} from '@react-navigation/stack';
+import {useMemo} from 'react';
+import useThemeStyles from '@hooks/useThemeStyles';
+import type {ThemeStyles} from '@src/styles';
+
+function useModalScreenOptions(getScreenOptions?: (styles: ThemeStyles) => StackNavigationOptions) {
```
Comment (roryabraham) [inline src/libs/Navigation/AppNavigator/ModalStackNavigators/useModalScreenOptions.ts:7]: Let's also use this hook in `createModalStackNavigator`

**Review by roryabraham — CHANGES_REQUESTED** [review]

File: src/libs/Navigation/AppNavigator/ModalStackNavigators/useModalScreenOptions.ts (hunk 2 above)
Comment (hayata-suenaga) [reply src/libs/Navigation/AppNavigator/ModalStackNavigators/useModalScreenOptions.ts:7]: fixed https://github.com/Expensify/App/pull/39084/commits/d5765d4a6c6a0cc4818b1ad3a49cf549089833cf 👍

**Review by roryabraham — CHANGES_REQUESTED** [review]
sorry I didn't notice it before, but it's a bit weird that we have a module structure like this:

```
Navigation/AppNavigator
----> ModalStackNavigtors.tsx
----| ModalStackNavigators
----------> WorkspaceSettingsModalStackNavigator.tsx
```

because it's not immediately obvious what you mean if you do `import ModalStackNavigators from @navigation/AppNavigator/ModalStackNavigator`

I think a directory structure that more closely matches the established patterns in this repo would be like this:

```
Navigation/AppNavigator
----| ModalStackNavigators
----------> index.tsx
----------> WorkspaceSettingsModalStackNavigator.tsx
```

**Comment by hayata-suenaga** [comment]
changed 😄 and tested again on web

https://github.com/Expensify/App/assets/98560306/ba38ce3b-36fa-44a6-8438-8bfc41108c35

**Review by roryabraham — APPROVED** [review]

**Pull request merged**